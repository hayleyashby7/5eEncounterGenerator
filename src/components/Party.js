import React from 'react';

const Party = ({ onSubmit }) => {
	return (
		<div>
			<form onSubmit={onSubmit}>
				<label htmlFor='char-num'>Characters (Max 10)</label>
				<input type='number' id='char-num' min='1' max='10' />
				<label htmlFor='level'>Level (1-20)</label>
				<input type='number' id='level' min='1' max='20' />
				<button name='partySubmit' type='submit'>
					Submit
				</button>
			</form>
		</div>
	);
};

export default Party;
