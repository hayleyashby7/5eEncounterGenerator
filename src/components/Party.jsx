import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import Difficulty from '../utils/difficulty';
import { Input, Select } from './Form';

function Party({ saveData }) {
    const {
        register,
        formState: { errors },
        handleSubmit,
    } = useForm({ mode: 'onBlur' });

    const onSubmit = (data) => {
        saveData(data);
    };

    const difficultyArray = [
        { value: null, label: 'Select a difficulty' },
        { value: Difficulty.Easy, label: 'Easy' },
        { value: Difficulty.Medium, label: 'Medium' },
        { value: Difficulty.Hard, label: 'Hard' },
        { value: Difficulty.Deadly, label: 'Deadly' },
    ];

    return (
        <form
            onSubmit={handleSubmit(onSubmit)}
            className='flex flex-col flex-wrap bg-stone-400 text-red-950 sm:flex-row '
        >
            <div className='flex flex-1 flex-col'>
                <Input
                    label='Number of characters'
                    name='numCharacters'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 10,
                        message: 'Max 10',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>

            <div className='flex flex-1 flex-col'>
                <Input
                    label='Level'
                    name='level'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 20,
                        message: 'Max 20',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col'>
                <Select
                    label='Difficulty'
                    name='difficulty'
                    options={difficultyArray}
                    register={register}
                    required
                    validate={(value) =>
                        Difficulty.difficultyType(value) || 'Must select a difficulty'
                    }
                />
                <ErrorMessage
                    errors={errors}
                    name='difficulty'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col justify-center'>
                <input
                    className='rounded-lg border-2 border-solid border-orange-100 bg-red-950 p-2 text-orange-100 hover:bg-red-900 hover:text-orange-200 focus:outline-none focus:ring-2 focus:ring-red-900/50'
                    name='partySubmit'
                    type='submit'
                />
            </div>
        </form>
    );
}

Party.propTypes = { saveData: PropTypes.func };
Party.defaultProps = { saveData: () => {} };

export default Party;
