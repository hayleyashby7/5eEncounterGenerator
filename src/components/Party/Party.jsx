import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import Difficulty from '../../utils/difficulty';
import Input from '../Form/Input';
import Select from '../Form/Select';
import Button from '../Form/Button';

function Party({ saveData }) {
    const {
        register,
        formState: { errors },
        handleSubmit,
    } = useForm({ mode: 'onBlur' });

    const onSubmit = (data) => {
        saveData(data);
    };

    const difficultyArray = [
        { value: null, label: 'Select a difficulty' },
        { value: Difficulty.Easy, label: 'Easy' },
        { value: Difficulty.Medium, label: 'Medium' },
        { value: Difficulty.Hard, label: 'Hard' },
        { value: Difficulty.Deadly, label: 'Deadly' },
    ];

    return (
        <form
            onSubmit={handleSubmit(onSubmit)}
            className='my-5 flex flex-col flex-wrap justify-end gap-4 rounded-lg bg-stone-400 p-5 text-red-950 lg:flex-row lg:justify-between'
        >
            <div className='flex flex-1 flex-col'>
                <Input
                    label='Number of characters'
                    name='numCharacters'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 10,
                        message: 'Max 10',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>

            <div className='flex flex-1 flex-col'>
                <Input
                    label='Level'
                    name='level'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 20,
                        message: 'Max 20',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='level'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col'>
                <Select
                    label='Difficulty'
                    name='difficulty'
                    options={difficultyArray}
                    register={register}
                    required
                    validate={(value) =>
                        Difficulty.difficultyType(value) || 'Must select a difficulty'
                    }
                />
                <ErrorMessage
                    errors={errors}
                    name='difficulty'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col justify-end'>
                <Button label='Generate' type='submit' name='generate' />
            </div>
        </form>
    );
}

Party.propTypes = { saveData: PropTypes.func };
Party.defaultProps = { saveData: () => {} };

export default Party;
