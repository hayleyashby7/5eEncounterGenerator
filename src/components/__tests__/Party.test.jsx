import React from 'react';
import { describe, it, expect } from 'vitest';
import { render, screen, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Party from '../Party';

describe('Party component', () => {
	it('renders the party composition form', () => {
		// Arrange
		const { container } = render(<Party />);

		//Assert
		expect(container).toMatchSnapshot();
	});

	it('form validation rejects number of characters outside of 1-10 values', async () => {
		// Arrange
		const user = userEvent.setup();

		render(<Party />);

		const input = screen.getByLabelText('Number of characters');

		// Act
		await user.type(input, '25');

		//Assert
		await act(() => {
			waitFor(() => {
				expect(input).toBeInvalid();
			});
		});
	});

	it('form validation rejects character level outside of 1-20 values', async () => {
		// Arrange
		const user = userEvent.setup();
		render(<Party />);
		const input = screen.getByLabelText('Level (1-20)');

		// Act
		await user.type(input, '68');

		//Assert
		await act(() => {
			waitFor(() => {
				expect(input).toBeInvalid();
			});
		});
	});
});
