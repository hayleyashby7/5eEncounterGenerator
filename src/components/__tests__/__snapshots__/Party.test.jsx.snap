// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Party component > renders the party composition form 1`] = `
<div>
  <form
    class="flex flex-col flex-wrap bg-stone-400 text-red-950 sm:flex-row "
  >
    <div
      class="flex flex-1 flex-col"
    >
      <label
        aria-labelledby="numCharacters"
        class="flex flex-1 flex-col"
        for="numCharacters"
      >
        Number of characters
        <input
          class="border-2 border-solid border-neutral-900 bg-orange-100 outline-none"
          id="numCharacters"
          name="numCharacters"
          type="number"
        />
      </label>
    </div>
    <div
      class="flex flex-1 flex-col"
    >
      <label
        aria-labelledby="level"
        class="flex flex-1 flex-col"
        for="level"
      >
        Level
        <input
          class="border-2 border-solid border-neutral-900 bg-orange-100 outline-none"
          id="level"
          name="level"
          type="number"
        />
      </label>
    </div>
    <div
      class="flex flex-1 flex-col"
    >
      <label
        aria-labelledby="difficulty"
        class="flex flex-1 flex-col"
        for="difficulty"
      >
        Difficulty
        <select
          class="border-2 border-solid border-neutral-900 bg-orange-100 outline-none"
          id="difficulty"
          name="difficulty"
        >
          <option>
            Select a difficulty
          </option>
          <option
            value="Easy"
          >
            Easy
          </option>
          <option
            value="Medium"
          >
            Medium
          </option>
          <option
            value="Hard"
          >
            Hard
          </option>
          <option
            value="Deadly"
          >
            Deadly
          </option>
        </select>
      </label>
    </div>
    <div
      class="flex flex-1 flex-col justify-center"
    >
      <input
        class="rounded-lg border-2 border-solid border-orange-100 bg-red-950 p-2 text-orange-100 hover:bg-red-900 hover:text-orange-200 focus:outline-none focus:ring-2 focus:ring-red-900/50"
        name="partySubmit"
        type="submit"
      />
    </div>
  </form>
</div>
`;
